beaker_kernel/lib/subkernels/base.py

```diff
@@ -128,3 +130,32 @@ def cleanup(self):
         if self.checkpoints_enabled:
             shutil.rmtree(self.storage_prefix, ignore_errors=True)
             self.checkpoints = []
+
+
+    @tool()
```

Comment:
Can you explain to me the thinking as to why it makes sense to define tools on using the @tool decorator on subkernel classes?
Previously, tools were only able to be defined as part of an agent and on a toolset class.
Keep in mind that the prompting for tools also takes in to consideration the parent class that an @tool is defined on.

Reply:
As I mentioned above in the PR:
> Agent inherits tools from specific subkernels. This means tools can now be specified for a language as opposed to a context.

[Toolsets are deprecated in archytas](https://github.com/jataware/archytas/blob/36c8ace7400b0b81abb24e7fa77a75289b6b0c5e/archytas/tool_utils.py#L41-L50). Any class can bundle together related functionality.

Currently, checkpointing only exists for Python so we only want to include the `run_code` tool when loading that specific subkernel. I've also in the past wanted to have some context-specific tools apply to all Julia contexts instead.

Reply:
Yes, I'm with you on defining tools that are defined depending on the kernel.

The thing that I'm getting at is that there are side-effects to how you are doing this, namely that it changes the prompt in ways I don't think you realize, and that there are alternative ways of doing what you want that would have less or no side-effects.
https://github.com/jataware/archytas/blob/main/archytas/tool_utils.py#L218

Namely: 
You can code this tool and similar tools as stand-alone functions and use the injections to get back to the agent/context for running code in the subkernel.

Reply:
All right, I solved this by collecting the tools from the object before passing them to the agent.

I implemented the current solution after looking into moving the tool out of the subkernel class. Converting was simple enough:
```python
@tool()
async def run_code(code: str, agent: AgentRef) -> str:
        """
        Execute code in the user's session. After execution,
        the state of the kernel will be rolled back to before this tool
        was used.

        This tool can be help answer questions about the kernel state. For
        example, a user may ask something about a dictionary `d` and using 
        run code with the `code` of `d.keys()`.

        This tool can also be used to double check if code will work before 
        returning it as a final answer.

        Note that this tool does not capture `stdout` AND only returns the
        results of the last expression evaluated.

        Args:
            code (str): Code to run directly in Jupyter.
        Returns:
            str: Result of the `expr`
        
        """
        checkpoint_index = await agent.context.subkernel.add_checkpoint()
        result = await agent.context.evaluate(code)
        await agent.context.subkernel.rollback(checkpoint_index)
        return str(result["return"])
```
It feels a bit funky to use `agent.context.subkernel` twice. We're separating the code pretty far from all the methods it's trying to use directly. 

Compounding the above, it makes thinking through enabling a bit trickier. I initially thought that the user could just pass `run_code` into context initialization if they wanted to use it. However, it gets tricky when they make a context that supports both a checkpointable and non-checkpointable kernel. Then, we'll need some other kind of logic to disable the tool when the non-checkpointable tool is used. We could add refs to the available tools in the subkernel class but that seems less ideal than what we have above. We could also just check using the language string although I think the current implementation (now that I've removed the class from the prompt) is succinct and intuitive. [Note this does not solve the more general questions about enabling](https://github.com/jataware/beaker-kernel/pull/47#discussion_r1594546342)